core/src/main/java/org/springframework/security/oauth2/server/authorization/InMemoryOAuth2AuthorizationService.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.server.authorization;
+
+import org.springframework.security.oauth2.core.OAuth2AccessToken;
+import org.springframework.util.Assert;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * In-memory implementation of {@link OAuth2AuthorizationService}.
+ *
+ * @author Krisztian Toth
+ */
+public class InMemoryOAuth2AuthorizationService implements OAuth2AuthorizationService {
+	private final Map<TokenWithType, OAuth2Authorization> tokenAuthorizationMap = new ConcurrentHashMap<>();
+
+	@Override
+	public void save(OAuth2Authorization authorization) {
+		Assert.notNull(authorization, "authorization cannot be null");
+
+		OAuth2AccessToken accessToken = authorization.getAccessToken();
+		String code = (String) authorization.getAttributes().get(TokenType.AUTHORIZATION_CODE.getValue());
+		if (accessToken != null) {
+			tokenAuthorizationMap.put(new TokenWithType(accessToken.getTokenValue(), TokenType.ACCESS_TOKEN),
+					authorization);
+		} else if (code != null) {
+			tokenAuthorizationMap.put(new TokenWithType(code, TokenType.AUTHORIZATION_CODE), authorization);
+		}
+	}
+
+	@Override
+	public OAuth2Authorization findByTokenAndTokenType(String token, TokenType tokenType) {
+		Assert.hasText(token, "token cannot be empty");
+		Assert.notNull(tokenType, "tokenType cannot be null");
+		return tokenAuthorizationMap.get(new TokenWithType(token, tokenType));
+	}
+
+	/**
+	 * Visible for testing.
+	 */
+	Map<TokenWithType, OAuth2Authorization> getTokenAuthorizationMap() {
```

Comment:
I'm not a fan of exposing a method strictly for testing. There usually is a way to test without exposing this method.

Reply:
Using constructor injection in `fd7ab91`, though I am still not sure if this is the best solution. Let me know. Also, which `List` implementation should be used, taking concurrency into account? I went with `CopyOnWriteArrayList` in the default constructor, but as far as I know its performance when writing to the list is really bad.